src/etc/nightshade.1
.\" Nightshade man page.
.\" FIX pull in intro from Info
.\"
.TH NIGHTSHADE 1 "May 01, 2009"
.AT 3
.SH NAME
nightshade \- a Lisp environment
.SH SYNOPSIS
.B nightshade
[input-files] [switch-list]
.SH DESCRIPTION
Start the Nightshade Lisp environment.  If
.I switch-list
is empty, then enter a read-eval-print loop using
stdin, stdout and stderr.  The optional
.I input-files
provide operands for certain switches.

The following switches are defined:
.
.TP 10n
.BR \-core " filename"
Specifies the suspended image (or `core file') to start up.  The default
is `lib/lisp.core'.
.TP
.BR \-edit
Enter the editor.
A file to edit may be specified by
placing the name of the file between the program name (usually `nightshade') and
the first switch.
.TP
.BR \-eval " expression"
Evaluates the specified Lisp
.I expression
during the start up sequence.  The value of the form will only be printed if
it is wrapped in a form that does output.
.TP
.BR \-init " filename"
Specifies the name of a file containing user customizations to be
loaded (usually the file is ~/.nightshade/lisp.lisp).
The loader
loads any compiled version of these files, else the source.
.TP
.BR \-einit " filename"
Similar to \-init, but specifies the name of the editor
init file (usually `~/.nightshade/ed.lisp'), which is loaded only
when the editor is started.
.TP
.BR \-noinit
Suppresses loading of the init file, and also prevents \-edit from loading the
editor init file.
.TP
.BR \-xoff
Instruct the editor to run in the terminal instead of as an X application.
.TP
.BR \-load " filename"
Loads the specified file before entering the read-eval-print loop.
.TP
.BR \-slave " editor-name"
Specifies that Lisp should start up as a
.I slave
Lisp and try to
connect to an editor Lisp.  The name of the editor to connect to must be
specified.  To find the editor's name, use the editor
`Accept Slave Connections' command.
.I editor-name
is of the form `machine-name:socket', where
.I machine-name
is the
internet host name for the machine and
.I socket
is the decimal number of the socket to connect to.
.PP
.\"
.SH ENVIRONMENT
.\"
.TP 10n
.BR NIGHTSHADELIB
The `lib/' directory holding `lisp.core' and other files used by Nightshade
at run-time.  Setting this variable overrides the standard location, which
is `/usr/local/lib/nightshade/lib'.
.TP
.BR TERMCAP
The terminal capabilities file.  Required by the editor when running under
a terminal.  Setting this variable overrides the standard location, which
is `/etc/termcap'.
.\" .TP
.\" .BR CMUCL_EMPTYFILE
.\" [SunOS only] If `df /tmp' shows `swap' as the filesystem for the `/tmp'
.\" directory, then you have a "tmpfs" filesystem.  In this case, you must setenv
.\" CMUCL_EMPTYFILE to point into a pathname on a non-TMPFS filesystem that can be
.\" used instead of `/tmp/empty'.
.\" .TP
.\" .BR XKEYSYMDB
.\" In order to use Motif (and the graphical debugger) with X servers from
.\" non-OSF vendors (like Sun) you may need to set the environment variable
.\" XKEYSYMDB to point to the file `lib/XKeysymDB'.  Otherwise, you will get many
.\" error messages every time a new connection is opened to the CMU CL motifd.
.\" This file is read by the X11R5 Xt in order to augment the keysym database with
.\" certain OSF vendor keysyms that Motif wants to use.
.
.PP
.\"
.SH EXAMPLES
.\"
.TP 10n
.BR nightshade " script"
Execute the code in script and then exit.
.\"
.TP
.BR nightshade " file -edit"
Open file in the editor.
.\" .TP
.PP
.\"
.SH FILES
.\"
The following pathnames are specified relative to the directory in which
Nightshade is installed.
.
.TP 10n
.\" .BR doc/*
.\" Various documentation files.
.\" .TP
.BR bin/nightshade
The lisp startup program.  This directory should be in PATH.
.TP
.BR lib/nightshade/lisp.core
The suspended Lisp image.
.TP
.BR lib/site-init.lisp
Site specific initialization.
.TP
.BR lib/hemlock11.*, lib/spell-dictionary.bin
Editor files.
.TP
.\" .BR lib/fonts/
.\" X11 fonts for Hemlock.
.\" .TP
.\" .BR lib/XKeysymDB
.\" Database of X Keysym names for Motif.
.\" .TP
.BR ~/.nightshade/lisp.lisp
User customization files loaded at lisp startup.
Init files can be compiled.
.TP
.BR ~/.nightshade/ed.lisp
User editor initialization file, loaded when the editor starts.
.PP
.\"
.SH ERRORS
.\"
Many.  Known errors are marked with "FIX" in the source.  etc:TODO is a
rough todo list.  Please send any feedback to matt@mundell.ukfsn.org.
.PP
.\"
.SH AUTHORS
.\"
Nightshade is based on the public domain portions of release 18c of CMUCL.
The authors of Nightshade include the authors of the original code at CMU,
contributers to the subsequent CMUCL internet-based project, and people who
have dedicated code to the public domain.
.\"
The "Authors" node in the documentation lists all authors.
.\"
.\"
.SH SEE ALSO
.\"
The full documentation, available in the editor via the `Doc' command
(bound to "control-h d").
.\"
The project home page online
.\"
.BR http://www.mundell.ukfsn.org/nightshade/
.\"
.\"
.\" FIX describe very basic ed intro exit
.\"
.SH COPYING
.\"
Nightshade is entirely Free Software.  Nightshade is most probably also
entirely in the public domain (further details at
http://www.mundell.ukfsn.org/nightshade/public.html).
.\" FIX more details
.\"


src/lisp/nightshade.c
/*
 * main() entry point for a stand alone lisp image.
 *
 * $Header: /home/CVS-cmucl/src/lisp/lisp.c,v 1.11.2.7 2000/10/24 17:46:15 dtc Exp $
 *
 */

#include <stdio.h>
#include <sys/types.h>
#include <stdlib.h>
#include <sys/file.h>
#include <sys/param.h>
#include <sys/stat.h>

#include "signal.h"

#include "nightshade.h"
#include "internals.h"
#include "alloc.h"
#include "vars.h"
#include "globals.h"
#include "os.h"
#include "interrupt.h"
#include "arch.h"
#include "gc.h"
#include "monitor.h"
#include "validate.h"
#if defined GENCGC
#include "gencgc.h"
#endif
#include "core.h"
#include "save.h"
#include "lispregs.h"

#ifdef irix
#include <string.h>
#include "interr.h"
#endif


/* SIGINT handler that invokes the monitor. */

static void sigint_handler(HANDLER_ARGS)
{
#if ( defined( __linux__ ) && defined( i386 ) )
  GET_CONTEXT
#endif

    SAVE_CONTEXT();

    printf("\nSIGINT hit at 0x%08lX\n", SC_PC(context));
    ldb_monitor();
}

/* Not static, because we want to be able to call it from lisp land. */
void sigint_init(void)
{
    install_handler(SIGINT, sigint_handler);
}


/* Noise to convert argv and argp into lists. */

static lispobj alloc_str_list(char *list[])
{
    lispobj result, newcons;
    struct cons *ptr;

    if (*list == NULL)
        result = NIL;
    else {
        result = newcons = alloc_cons(alloc_string(*list++), NIL);

        while (*list != NULL) {
            ptr = (struct cons *)PTR(newcons);
            newcons = alloc_cons(alloc_string(*list++), NIL);
            ptr->cdr = newcons;
        }
    }

    return result;
}


/* And here be main. */

void main(int argc, char *argv[], char *envp[])
{
    char *arg, **argptr;
    char *core = NULL, *default_core;
    boolean monitor;
    lispobj initial_function;

#ifdef MACH
    mach_init();
#endif
#if defined(SVR4) || defined(__linux__)
    tzset();
#endif

    set_lossage_handler(ldb_monitor);

    monitor = FALSE;
#ifdef DEFAULT_DYNAMIC_SPACE_SIZE
    dynamic_space_size = DEFAULT_DYNAMIC_SPACE_SIZE;
#else
    dynamic_space_size = DYNAMIC_SPACE_SIZE;
#endif

    argptr = argv;
    while ((arg = *++argptr) != NULL)
      {
        if (strcmp(arg, "-core") == 0)
	  {
            if (core != NULL)
	      {
                fprintf(stderr, "can only specify one core file.\n");
                exit(1);
	      }
            core = *++argptr;
            if (core == NULL)
	      {
		fprintf(stderr, "-core must be followed by the name of the core file to use.\n");
                exit(1);
	      }
	  }
        else if (strcmp(arg, "-dynamic-space-size") == 0)
	  {
            char *str = *++argptr;
            if (str == NULL)
	      {
                fprintf(stderr, "-dynamic-space-size must be followed by the size to use in MBytes.\n");
                exit(1);
	      }
	    dynamic_space_size = atoi(str) * 1024 * 1024;
	    if (dynamic_space_size > DYNAMIC_SPACE_SIZE)
	      {
                fprintf(stderr, "-dynamic-space-size must be at most %d MBytes.\n",
			DYNAMIC_SPACE_SIZE / (1024 * 1024));
                exit(1);
	      }
	  }
	else if (strcmp(arg, "-monitor") == 0)
	  {
	    monitor = TRUE;
	  }
      }

    default_core = arch_init();
    if (default_core == NULL)
	default_core = "lisp.core";

    if (core == NULL) {
#ifdef MACH
	extern char *getenv(char *var);
#endif
	static char buf[MAXPATHLEN];
	char *lib = getenv("NIGHTSHADELIB");

	if (lib != NULL) {
	    char *dst;
	    struct stat statbuf;

	    do {
		dst = buf;
		while (*lib != '\0' && *lib != ':')
		    *dst++ = *lib++;
		if (dst != buf && dst[-1] != '/')
		    *dst++ = '/';
		strcpy(dst, default_core);
		if (stat(buf, &statbuf) == 0) {
		    core = buf;
		    break;
		}
	    } while (*lib++ == ':');
	}
	if (core == NULL) {
	    /* Note: the /usr/misc/.nightshade/lib/ default path is also wired
	       into the lisp code in .../code/save.lisp. */
#ifdef MACH
	    strcpy(buf, "/usr/misc/.nightshade/lib/");
#else
	    strcpy(buf, "/usr/local/lib/nightshade/");
#endif
	    strcat(buf, default_core);
	    core = buf;
	}
    }

    os_init();
    validate();
    gc_init();

    /* This is the first use of malloc() and must come after the
     * static memory layout is mmapped to avoid conflicts with possible
     * use of mmap() by malloc().
     */
    define_var("nil", NIL, TRUE);
    define_var("t"  ,   T, TRUE);

    globals_init();

    initial_function = load_core_file(core);


#if defined GENCGC
    gencgc_pickup_dynamic();
#else
#if defined WANT_CGC && defined X86_CGC_ACTIVE_P
    {
      extern int use_cgc_p;
      lispobj x = SymbolValue(X86_CGC_ACTIVE_P);
      if(x != type_UnboundMarker && x != NIL)
	use_cgc_p = 1;		/* enable allocator */
    }
#endif
#endif

#ifdef BINDING_STACK_POINTER
    SetSymbolValue(BINDING_STACK_POINTER, (lispobj)binding_stack);
#endif
#if defined INTERNAL_GC_TRIGGER && !defined i386
    SetSymbolValue(INTERNAL_GC_TRIGGER, make_fixnum(-1));
#endif

    interrupt_init();

    arch_install_interrupt_handlers();
    os_install_interrupt_handlers();

#ifdef PSEUDO_ATOMIC_ATOMIC
    /* Turn on pseudo atomic for when we call into lisp. */
    SetSymbolValue(PSEUDO_ATOMIC_ATOMIC, make_fixnum(1));
    SetSymbolValue(PSEUDO_ATOMIC_INTERRUPTED, make_fixnum(0));
#endif

    /* Convert the argv and envp to something Lisp can grok. */
    SetSymbolValue(LISP_COMMAND_LINE_LIST, alloc_str_list(argv));
    SetSymbolValue(LISP_ENVIRONMENT_LIST, alloc_str_list(envp));

    // FIX must kernel.core have -batch?
    /*
     * Parse the command line again, picking up values that override
     * those loaded from the core.
     */

    argptr = argv;
    while ((arg = *++argptr) != NULL)
      {
	if (strcmp(arg, "-batch") == 0)
	  SetSymbolValue(BATCH_MODE, T);
      }

    /*
     * Pick off sigint until the lisp system gets far enough along to
     * install it's own.
     */
    sigint_init();

    if (monitor)
	while (1)
	    ldb_monitor();
    else {
	funcall0(initial_function);
	printf("Initial function returned?\n");
	exit(1);
    }
}
